src/codegen/compile.ml

```diff
@@ -7277,27 +7201,29 @@ module Serialization = struct
 
         (* If it is a reference, temporarily set the read buffer to that place *)
         get_is_ref ^^
-        G.if0 begin
+        E.if0 begin
           let (set_offset, get_offset) = new_local env "offset" in
-          ReadBuf.read_word32 env get_data_buf ^^ set_offset ^^
+          ReadBuf.read_signed_word32 env get_data_buf ^^ 
+          set_offset ^^
           (* A sanity check *)
-          get_offset ^^ compile_unboxed_const 0l ^^
-          G.i (Compare (Wasm.Values.I32 I32Op.LtS)) ^^
+          get_offset ^^ compile_unboxed_const 0L ^^
+          compile_comparison I64Op.LtS ^^
           E.else_trap_with env "Odd offset" ^^
 
           ReadBuf.get_ptr get_data_buf ^^ set_cur ^^
-          ReadBuf.advance get_data_buf (get_offset ^^ compile_add_const (-4l))
+          ReadBuf.advance get_data_buf (get_offset ^^ compile_add_const (-4L))
         end G.nop ^^
 
         (* Remember location of ptr *)
         ReadBuf.get_ptr get_data_buf ^^ set_memo ^^
         (* Did we decode this already? *)
-        ReadBuf.read_word32 env get_data_buf ^^ set_result ^^
-        get_result ^^ compile_eq_const 0l ^^
-        G.if0 begin
+        ReadBuf.read_signed_word32 env get_data_buf ^^ 
```

Comment:
Note for myself: I believe this is wrong because it reads a 32-bit narrowed heap pointer. This is also related to the problem that the Candid destabilization cannot deserialize objects that are allocated beyond 4GB which could be possible because the 64-bit representation is larger than on 32-bit. In other words, a 32-bit 4GB stabilized data may inflate to more than 4GB on 64-bit and then the 32-pointer storage for aliases in Candid buffer does no longer suffice.

Reply:
Yeah, I was abit suspicious about this new function, but I think it's only used in 2 place and might be ok, but worth checking again. certainly didn't seem necessary.  Or maybe I'm confusing it with some other read_signed_word32

Reply:
**Update**: This is now fixed by using pointer compression in the memo, see comment above.